File: src/main/java/com/jgji/selfassessment/global/config/security/token/JwtProvider.java
package com.jgji.selfassessment.global.config.security.token;

import com.jgji.selfassessment.global.config.security.token.properties.JwtProperties;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.security.Key;
import java.util.Date;

@Slf4j
@Component
class JwtProvider implements TokenProvider {

    private final JwtProperties jwtProperties;
    private final Key key;
    private final long expirationMs;

    public JwtProvider(JwtProperties jwtProperties) {
        this.jwtProperties = jwtProperties;
        this.key = Keys.hmacShaKeyFor(jwtProperties.getSecretKey().getBytes());
        this.expirationMs = jwtProperties.getExpirationMs();
    }

    public String generateJwtToken(String username) {
        return Jwts.builder()
            .setSubject(username)
            .setIssuedAt(new Date())
            .setExpiration(new Date((new Date()).getTime() + expirationMs))
            .signWith(key, SignatureAlgorithm.HS512)
            .compact();
    }

    public String getUsernameFromJwtToken(String token) {
        return Jwts.parser()
            .setSigningKey(key)
            .build()
            .parseClaimsJws(token)
            .getBody()
            .getSubject();
    }

    public boolean validateJwtToken(String authToken) {
        try {
            Jwts.parser()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(authToken);
            return true;
        } catch (JwtException e) {
            log.warn("토큰 검증 실패 : {}", authToken);
        }
        return false;
    }
}

File: src/main/kotlin/com/jgji/selfassessment/global/infra/token/JwtProvider.kt
package com.jgji.selfassessment.global.infra.token

import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtProvider : TokenProvider {

    private val jwtSecret = "mySecretKey" // 비밀키, 실제 환경에서는 외부에서 주입받아야 함
    private val jwtExpirationMs = 3600000L // 토큰 유효 시간 (예: 1시간)
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(Decoders.BASE64URL.decode(jwtSecret))

    fun generateJwt(username: String): String {
        return Jwts.builder()
            .subject(username)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + jwtExpirationMs))
            .signWith(secretKey, Jwts.SIG.HS512)
            .compact();
    }

    fun getUsernameFromJwt(jwt: String): String {
        return Jwts.parser()
            .verifyWith(secretKey)
            .build()
            .parseSignedClaims(jwt)
            .payload
            .subject
    }

    fun validateJwt(authToken: String): Boolean {
        try {
            Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(authToken)
            return true;
        } catch (e: JwtException) {
            e.printStackTrace()
        }

        return false;
    }

}